File: lib/psychometrics.js
// ─────────────────────────────────────────────────────────────────────────
// Psychometrics core — EIE Phase 2 (Measurement Engine)
// Turns an ordered BARS rating into a Bayesian posterior over a latent ability
// θ using Samejima's Graded Response Model (GRM) with EAP estimation over a
// fixed θ-grid. Outputs a point estimate, a credible interval (honest
// uncertainty), a norm-percentile, and marginal reliability.
//
// Why this and not "level×20": a point score hides how well we actually
// measured the person. A confident, well-evidenced L4 and a hand-wavy L4 must
// NOT look identical — here the second gets a wide credible interval. This is
// the single biggest scientific upgrade over every "AI scoring" competitor.
//
// See EVALUATION_ENGINE.md §4. Real multi-rater G-theory reliability lands in
// Phase 3; here reliability is the IRT *marginal* reliability (variance
// reduction vs the prior), which is the correct single-rater analogue.
// ─────────────────────────────────────────────────────────────────────────

// Standard-normal pdf / cdf (Abramowitz & Stegun 7.1.26 erf approximation).
function normalPdf(z) {
  return Math.exp(-0.5 * z * z) / Math.sqrt(2 * Math.PI)
}
export function normalCdf(z) {
  const t = 1 / (1 + 0.2316419 * Math.abs(z))
  const d = 0.3989423 * Math.exp(-z * z / 2)
  let p = d * t * (0.3193815 + t * (-0.3565638 + t * (1.781478 + t * (-1.821256 + t * 1.330274))))
  return z > 0 ? 1 - p : p
}

const logistic = (x) => 1 / (1 + Math.exp(-x))
const clamp = (n, lo, hi) => Math.max(lo, Math.min(hi, n))

// θ grid: -4..4 step 0.1 (81 points). Fixed → deterministic, fast, no sampler.
const GRID = Array.from({ length: 81 }, (_, i) => -4 + i * 0.1)

// Default GRM parameters (theory-set; refit from data in Phase 6).
//  - discrimination a: how sharply the anchors separate ability.
//  - thresholds: the 4 category boundaries between the 5 BARS levels on the θ
//    scale. Symmetric around 0 so L3 ≈ population median, L5 ≈ +1.8 SD.
export const DEFAULT_A = 2.0
export const DEFAULT_THRESHOLDS = [-1.5, -0.5, 0.5, 1.5]

// P(response == level | θ) under the GRM.
function categoryProb(level, theta, a, thr) {
  const pge = (k) => (k <= 1 ? 1 : k >= 6 ? 0 : logistic(a * (theta - thr[k - 2])))
  return clamp(pge(level) - pge(level + 1), 1e-9, 1)
}

// Bayesian posterior over θ for a single observed BARS `level`.
//  weight ∈ (0,1] = rater confidence → tempers the likelihood (low confidence
//  ⇒ near-flat likelihood ⇒ posterior ≈ prior ⇒ wide credible interval).
// `weight` is the effective number of (confidence-scaled) observations of θ —
// under local independence, multiple cited behavioral episodes each inform θ,
// so weight can exceed 1 (more evidence ⇒ sharper posterior ⇒ tighter CI).
export function grmPosterior({ level, weight = 1, a = DEFAULT_A, thresholds = DEFAULT_THRESHOLDS, priorMean = 0, priorSD = 1 }) {
  const w = clamp(weight, 0.05, 3.5)
  const unnorm = GRID.map((t) => {
    const prior = normalPdf((t - priorMean) / priorSD)
    const lik = categoryProb(level, t, a, thresholds)
    return prior * Math.pow(lik, w)
  })
  const Z = unnorm.reduce((s, x) => s + x, 0) || 1
  const post = unnorm.map((x) => x / Z)

  const theta = GRID.reduce((s, t, i) => s + t * post[i], 0)
  const variance = GRID.reduce((s, t, i) => s + (t - theta) ** 2 * post[i], 0)
  const se = Math.sqrt(variance)

  // 90% credible interval from the posterior CDF over the grid.
  const ci90 = [quantile(post, 0.05), quantile(post, 0.95)]

  // IRT marginal reliability: 1 − (posterior var / prior var).
  const reliability = clamp(1 - variance / (priorSD * priorSD), 0, 1)

  return {
    theta: round(theta, 2),
    se: round(se, 2),
    ci90: [round(ci90[0], 2), round(ci90[1], 2)],
    reliability: round(reliability, 2),
    percentile: Math.round(normalCdf(theta) * 100), // norm-population percentile
  }
}

// Multi-observation posterior — the Phase 3 core. Each observation is one
// rater's BARS level (with its own weight). Under local independence the
// likelihood is the product over observations, so AGREEING raters sharpen the
// posterior (tight CI, high reliability) and DISAGREEING raters flatten it
// (wide CI, low reliability). Rater disagreement thus becomes honest uncertainty
// automatically — no ad-hoc penalty needed. See EVALUATION_ENGINE.md §4.6.
export function grmPosteriorMulti({ observations, a = DEFAULT_A, thresholds = DEFAULT_THRESHOLDS, priorMean = 0, priorSD = 1 }) {
  const obs = (observations || []).filter((o) => Number.isFinite(o.level))
  if (!obs.length) return null
  const unnorm = GRID.map((t) => {
    let logp = Math.log(normalPdf((t - priorMean) / priorSD) + 1e-300)
    for (const o of obs) {
      const w = clamp(o.weight ?? 1, 0.02, 3.5)
      logp += w * Math.log(categoryProb(o.level, t, a, thresholds))
    }
    return Math.exp(logp)
  })
  const Z = unnorm.reduce((s, x) => s + x, 0) || 1
  const post = unnorm.map((x) => x / Z)

  const theta = GRID.reduce((s, t, i) => s + t * post[i], 0)
  const variance = GRID.reduce((s, t, i) => s + (t - theta) ** 2 * post[i], 0)
  const se = Math.sqrt(variance)
  const ci90 = [quantile(post, 0.05), quantile(post, 0.95)]
  const reliability = clamp(1 - variance / (priorSD * priorSD), 0, 1)
  return {
    theta: round(theta, 2), se: round(se, 2),
    ci90: [round(ci90[0], 2), round(ci90[1], 2)],
    reliability: round(reliability, 2),
    percentile: Math.round(normalCdf(theta) * 100),
  }
}

// Exact-agreement-style rater agreement for a set of BARS levels (0..1):
// 1 − (mean absolute deviation from the modal level / max possible deviation).
export function raterAgreement(levels) {
  const xs = (levels || []).filter((n) => Number.isFinite(n))
  if (xs.length < 2) return null
  const mean = xs.reduce((s, x) => s + x, 0) / xs.length
  const mad = xs.reduce((s, x) => s + Math.abs(x - mean), 0) / xs.length
  return round(clamp(1 - mad / 2, 0, 1), 2) // 2 = half the 1..5 range
}

// Two-way (competency × rater) variance decomposition on a level matrix →
// a G-theory dependability for the candidate's PROFILE against rater swaps.
// matrix: rows = competencies, cols = raters (equal length rows).
export function profileDependability(matrix) {
  const rows = (matrix || []).filter((r) => Array.isArray(r) && r.length)
  const nC = rows.length
  const nR = rows[0]?.length || 0
  if (nC < 2 || nR < 2) return null
  const grand = rows.flat().reduce((s, x) => s + x, 0) / (nC * nR)
  const compMeans = rows.map((r) => r.reduce((s, x) => s + x, 0) / nR)
  const raterMeans = Array.from({ length: nR }, (_, j) => rows.reduce((s, r) => s + r[j], 0) / nC)

  let ssC = 0, ssR = 0, ssE = 0
  for (let i = 0; i < nC; i++) {
    ssC += nR * (compMeans[i] - grand) ** 2
    for (let j = 0; j < nR; j++) {
      const resid = rows[i][j] - compMeans[i] - raterMeans[j] + grand
      ssE += resid * resid
    }
  }
  for (let j = 0; j < nR; j++) ssR += nC * (raterMeans[j] - grand) ** 2
  const vComp = Math.max(0, ssC / (nC - 1) - ssE / ((nC - 1) * (nR - 1))) // simplified EMS
  const vRater = Math.max(0, ssR / (nR - 1))
  const vResid = Math.max(1e-6, ssE / ((nC - 1) * (nR - 1)))
  // Dependability is undefined when there is no between-competency (true-score)
  // variance — a flat profile means "nothing to differentiate", NOT "unreliable".
  const dependability = vComp < 1e-3 ? null : round(clamp(vComp / (vComp + (vRater + vResid) / nR), 0, 1), 2)
  return { vComp: round(vComp, 3), vRater: round(vRater, 3), vResid: round(vResid, 3), dependability }
}

function quantile(post, q) {
  let cum = 0
  for (let i = 0; i < GRID.length; i++) {
    cum += post[i]
    if (cum >= q) return GRID[i]
  }
  return GRID[GRID.length - 1]
}

function round(n, d) {
  const f = 10 ** d
  return Math.round(n * f) / f
}

// Combine per-competency posteriors into a weighted composite θ with correct
// error propagation. weights need not sum to 1 (normalized internally).
export function compositePosterior(items) {
  // items: [{ theta, se, weight }]
  const measured = items.filter((x) => Number.isFinite(x.theta) && x.weight > 0)
  if (!measured.length) return null
  const W = measured.reduce((s, x) => s + x.weight, 0)
  const theta = measured.reduce((s, x) => s + (x.weight / W) * x.theta, 0)
  const variance = measured.reduce((s, x) => s + ((x.weight / W) ** 2) * (x.se * x.se), 0)
  const se = Math.sqrt(variance)
  return {
    theta: round(theta, 2),
    se: round(se, 2),
    ci90: [round(theta - 1.645 * se, 2), round(theta + 1.645 * se, 2)],
    percentile: Math.round(normalCdf(theta) * 100),
  }
}

// Seniority base-rate prior mean (weak, non-zero information). Senior norm = 0.
export function priorMeanForSeniority(seniority) {
  return { ic3: -0.3, senior: 0, staff: 0.2, exec: 0.3 }[seniority] ?? 0
}

File: backend/domain/evaluation/psychometrics.py
"""Psychometrics core — EIE Measurement Engine (ported from lib/psychometrics.js).

Turns an ordered BARS rating into a Bayesian posterior over a latent ability θ
using Samejima's Graded Response Model (GRM) with EAP estimation over a fixed
θ-grid. Outputs a point estimate, a credible interval (honest uncertainty), a
norm-percentile, and marginal reliability.

Why this and not "level×20": a point score hides how well we actually measured
the person. A confident, well-evidenced L4 and a hand-wavy L4 must NOT look
identical — the second gets a wide credible interval. This is a faithful port of
the JS engine; the math is byte-for-byte equivalent (including JS Math.round
half-up rounding), verified against the same regression assertions.
"""

from __future__ import annotations

import math
from dataclasses import dataclass


# ── JS-parity helpers ────────────────────────────────────────────────────────
def _js_round(n: float, d: int = 0) -> float:
    """Match JavaScript Math.round(n*10^d)/10^d — half rounds toward +Infinity."""
    f = 10 ** d
    r = math.floor(n * f + 0.5) / f
    return int(r) if d == 0 else r


def normal_pdf(z: float) -> float:
    return math.exp(-0.5 * z * z) / math.sqrt(2 * math.pi)


def normal_cdf(z: float) -> float:
    """Standard-normal CDF (Abramowitz & Stegun 7.1.26)."""
    t = 1 / (1 + 0.2316419 * abs(z))
    d = 0.3989423 * math.exp(-z * z / 2)
    p = d * t * (0.3193815 + t * (-0.3565638 + t * (1.781478 + t * (-1.821256 + t * 1.330274))))
    return 1 - p if z > 0 else p


def _logistic(x: float) -> float:
    return 1 / (1 + math.exp(-x))


def _clamp(n: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, n))


# θ grid: -4..4 step 0.1 (81 points). Fixed → deterministic, fast, no sampler.
GRID: list[float] = [-4 + i * 0.1 for i in range(81)]

# Default GRM parameters (theory-set; refit from data in the calibration phase).
DEFAULT_A = 2.0
DEFAULT_THRESHOLDS = [-1.5, -0.5, 0.5, 1.5]


def _category_prob(level: int, theta: float, a: float, thr: list[float]) -> float:
    """P(response == level | θ) under the GRM."""
    def pge(k: int) -> float:
        if k <= 1:
            return 1.0
        if k >= 6:
            return 0.0
        return _logistic(a * (theta - thr[k - 2]))
    return _clamp(pge(level) - pge(level + 1), 1e-9, 1)


def _quantile(post: list[float], q: float) -> float:
    cum = 0.0
    for i in range(len(GRID)):
        cum += post[i]
        if cum >= q:
            return GRID[i]
    return GRID[-1]


@dataclass
class Posterior:
    theta: float
    se: float
    ci90: tuple[float, float]
    percentile: int
    reliability: float | None = None


def grm_posterior(
    level: int,
    weight: float = 1.0,
    a: float = DEFAULT_A,
    thresholds: list[float] | None = None,
    prior_mean: float = 0.0,
    prior_sd: float = 1.0,
) -> Posterior:
    """Bayesian posterior over θ for a single observed BARS `level`.

    `weight` = effective number of confidence-scaled observations; >1 sharpens
    the posterior (tighter CI), low confidence flattens it toward the prior.
    """
    thr = thresholds if thresholds is not None else DEFAULT_THRESHOLDS
    w = _clamp(weight, 0.05, 3.5)
    unnorm = [
        normal_pdf((t - prior_mean) / prior_sd) * (_category_prob(level, t, a, thr) ** w)
        for t in GRID
    ]
    z = sum(unnorm) or 1.0
    post = [x / z for x in unnorm]

    theta = sum(t * post[i] for i, t in enumerate(GRID))
    variance = sum((t - theta) ** 2 * post[i] for i, t in enumerate(GRID))
    se = math.sqrt(variance)
    ci90 = (_quantile(post, 0.05), _quantile(post, 0.95))
    reliability = _clamp(1 - variance / (prior_sd * prior_sd), 0, 1)

    return Posterior(
        theta=_js_round(theta, 2),
        se=_js_round(se, 2),
        ci90=(_js_round(ci90[0], 2), _js_round(ci90[1], 2)),
        reliability=_js_round(reliability, 2),
        percentile=_js_round(normal_cdf(theta) * 100),
    )


def grm_posterior_multi(
    observations: list[dict],
    a: float = DEFAULT_A,
    thresholds: list[float] | None = None,
    prior_mean: float = 0.0,
    prior_sd: float = 1.0,
) -> Posterior | None:
    """Multi-rater posterior. Agreeing raters sharpen the posterior; disagreeing
    raters flatten it — rater disagreement becomes honest uncertainty for free.

    observations: list of {"level": int, "weight": float}.
    """
    thr = thresholds if thresholds is not None else DEFAULT_THRESHOLDS
    obs = [o for o in (observations or []) if isinstance(o.get("level"), (int, float)) and math.isfinite(o["level"])]
    if not obs:
        return None
    unnorm: list[float] = []
    for t in GRID:
        logp = math.log(normal_pdf((t - prior_mean) / prior_sd) + 1e-300)
        for o in obs:
            w = _clamp(o.get("weight", 1) if o.get("weight") is not None else 1, 0.02, 3.5)
            logp += w * math.log(_category_prob(int(o["level"]), t, a, thr))
        unnorm.append(math.exp(logp))
    z = sum(unnorm) or 1.0
    post = [x / z for x in unnorm]

    theta = sum(t * post[i] for i, t in enumerate(GRID))
    variance = sum((t - theta) ** 2 * post[i] for i, t in enumerate(GRID))
    se = math.sqrt(variance)
    ci90 = (_quantile(post, 0.05), _quantile(post, 0.95))
    reliability = _clamp(1 - variance / (prior_sd * prior_sd), 0, 1)
    return Posterior(
        theta=_js_round(theta, 2),
        se=_js_round(se, 2),
        ci90=(_js_round(ci90[0], 2), _js_round(ci90[1], 2)),
        reliability=_js_round(reliability, 2),
        percentile=_js_round(normal_cdf(theta) * 100),
    )


def rater_agreement(levels: list[float]) -> float | None:
    """Exact-agreement-style rater agreement for BARS levels (0..1)."""
    xs = [x for x in (levels or []) if isinstance(x, (int, float)) and math.isfinite(x)]
    if len(xs) < 2:
        return None
    mean = sum(xs) / len(xs)
    mad = sum(abs(x - mean) for x in xs) / len(xs)
    return _js_round(_clamp(1 - mad / 2, 0, 1), 2)


@dataclass
class Dependability:
    v_comp: float
    v_rater: float
    v_resid: float
    dependability: float | None


def profile_dependability(matrix: list[list[float]]) -> Dependability | None:
    """Two-way (competency × rater) variance decomposition → G-theory
    dependability of the candidate's PROFILE against rater swaps.
    Returns None dependability when the profile is flat (nothing to differentiate).
    """
    rows = [r for r in (matrix or []) if isinstance(r, list) and len(r)]
    n_c = len(rows)
    n_r = len(rows[0]) if rows else 0
    if n_c < 2 or n_r < 2:
        return None
    flat = [x for r in rows for x in r]
    grand = sum(flat) / (n_c * n_r)
    comp_means = [sum(r) / n_r for r in rows]
    rater_means = [sum(rows[i][j] for i in range(n_c)) / n_c for j in range(n_r)]

    ss_c = ss_r = ss_e = 0.0
    for i in range(n_c):
        ss_c += n_r * (comp_means[i] - grand) ** 2
        for j in range(n_r):
            resid = rows[i][j] - comp_means[i] - rater_means[j] + grand
            ss_e += resid * resid
    for j in range(n_r):
        ss_r += n_c * (rater_means[j] - grand) ** 2
    v_comp = max(0.0, ss_c / (n_c - 1) - ss_e / ((n_c - 1) * (n_r - 1)))
    v_rater = max(0.0, ss_r / (n_r - 1))
    v_resid = max(1e-6, ss_e / ((n_c - 1) * (n_r - 1)))
    dependability = (
        None if v_comp < 1e-3
        else _js_round(_clamp(v_comp / (v_comp + (v_rater + v_resid) / n_r), 0, 1), 2)
    )
    return Dependability(
        v_comp=_js_round(v_comp, 3),
        v_rater=_js_round(v_rater, 3),
        v_resid=_js_round(v_resid, 3),
        dependability=dependability,
    )


def composite_posterior(items: list[dict]) -> Posterior | None:
    """Combine per-competency posteriors into a weighted composite θ with correct
    error propagation. items: [{"theta", "se", "weight"}]. Weights normalized.
    """
    measured = [
        x for x in items
        if isinstance(x.get("theta"), (int, float)) and math.isfinite(x["theta"]) and x.get("weight", 0) > 0
    ]
    if not measured:
        return None
    total_w = sum(x["weight"] for x in measured)
    theta = sum((x["weight"] / total_w) * x["theta"] for x in measured)
    variance = sum(((x["weight"] / total_w) ** 2) * (x["se"] * x["se"]) for x in measured)
    se = math.sqrt(variance)
    return Posterior(
        theta=_js_round(theta, 2),
        se=_js_round(se, 2),
        ci90=(_js_round(theta - 1.645 * se, 2), _js_round(theta + 1.645 * se, 2)),
        percentile=_js_round(normal_cdf(theta) * 100),
    )


def prior_mean_for_seniority(seniority: str) -> float:
    """Seniority base-rate prior mean (weak, non-zero information)."""
    return {"ic3": -0.3, "senior": 0.0, "staff": 0.2, "exec": 0.3}.get(seniority, 0.0)
